bpo-43795: Add a test for Stable ABI symbol availability using ctypes

encukou opened the pull request:
This is a cross-platform check that the symbols are actually
exported in the ABI, not e.g. hidden in a macro.


<!-- issue-number: [bpo-43795](https://bugs.python.org/issue43795) -->
https://bugs.python.org/issue43795
<!-- /issue-number -->

encukou committed:
Add a test for Stable ABI symbol availability using ctypes

This is a cross-platform check that the symbols are actually
exported in the ABI, not e.g. hidden in a macro.

erlend-aasland commented on Tools/scripts/stable_abi.py:
```diff
@@ -259,6 +259,37 @@ def gen_doc_annotations(manifest, args, outfile):
+    """Generate/check the ctypes-based test for exported symbols"""
+    write = partial(print, file=outfile)
+    write(textwrap.dedent('''
+        # Generated by Tools/scripts/stable_abi.py
```
Should this be checked by the _"Check if generated files are up to date"_ CI?

encukou replied:
Yes. I made the classic mistake, renaming without running the tests :/

encukou committed:
Rename the test class

encukou committed:
Remove reference to Py_FrozenMain

encukou committed:
Skip test when _ctypes is not available

encukou committed:
Merge main branch

encukou committed:
Regenerate the test file

encukou committed:
Skip the test for PyModule_Create2 & PyModule_FromDefAndSpec2

These aren't exported on some platforms.
All the others should still be tested, though, to make sure
they don't disappear.
